modules/standard/IO.chpl

```diff
@@ -6761,119 +6876,120 @@ proc readf(fmt:string):bool throws {
 
    this function will skip to (but leave unread) the comma after
    the first field value.
-
-   :arg error: optional argument to capture an error code. If this argument
-               is not provided and an error is encountered, this function
-               will halt with an error message.
  */
-proc channel.skipField(out error:syserr) {
+proc channel.skipField() throws {
+  var err:syserr = ENOERR;
   on this.home {
     try! this.lock();
     var st = this.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
     if st == QIO_AGGREGATE_FORMAT_JSON {
-      error = qio_channel_skip_json_field(false, _channel_internal);
+      err = qio_channel_skip_json_field(false, _channel_internal);
     } else {
-      error = ENOTSUP;
+      err = ENOTSUP;
     }
     this.unlock();
   }
+  if err then try this._ch_ioerror(err, "in skipField");
 }
+
 pragma "no doc"
-proc channel.skipField() throws {
-  var err:syserr;
-  this.skipField(err);
-  if err then try this._ch_ioerror(err, "in skipField");
+proc channel.skipField(out error:syserr) {
+  compilerWarning("This version of channel.skipField() is deprecated; " +
+                  "please switch to a throwing version");
+  try {
+    this.skipField();
+  } catch e: SystemError {
+    error = e.err;
+  } catch {
+    error = EINVAL;
+  }
+}
+
+
+// This function is no longer available.
+pragma "no doc"
+proc format(fmt:string, args ...?k):string {
+  compilerError("use string.format(args ...) not format(fmt, args ...)");
 }
 
+/*
+
+  Return a new string consisting of values formatted according to a
+  format string.  See :ref:`about-io-formatted-io`.
 
-private inline proc chpl_do_format(fmt:string, args ...?k, out error:syserr):string {
+  :arg this: the format string
+  :arg args: the arguments to format
+  :returns: the resulting string
+ */
+proc string.format(args ...?k): string throws {
+  try {
+    return chpl_do_format(this, (...args));
+  } catch e: SystemError {
+    try ioerror(e.err, "in string.format");
+  } catch {
+    try ioerror(EINVAL:syserr, "in string.format");
+  }
+  return "";
+}
 
+private inline proc chpl_do_format(fmt:string, args ...?k): string throws {
   // Open a memory buffer to store the result
-  var f = openmem(error=error);
+  var f = try openmem();
   defer {
-    var closeError:syserr;
-    f.close(error=closeError);
-    // Propagate an error on closing only if there wasn't an original error.
-    if !error then
-      error = closeError;
+    try {
+      f.close();
+    } catch {
+      // ignore close errors
```

Comment:
I'd prefer it if this code still reported an error on close if one occurred.
It could do that by calling close explicitly outside of the defer block, and then redundantly calling it in the defer block. Or replacing the defer block with a try / catch where the error handling case closes the files and re-throws.

Reply:
Should this function be allowed to return a valid output when any of the `close()`s throw an error? The function seems to have allowed that before, so if we need to maintain that, the patterns you describe won't work.

Reply:
I think it shouldn't return output if any of the closes throw an error.

Reply:
Alright, I've updated this function to throw on an close error, supposing another error did not occur.